File: frontend/src/lib/parada.ts
// Que parte de una condicion de parada mide Rosa sola. Espejo de rosa/parada.py:
// misma regla en los dos lados, para que la interfaz lo muestre al escribirla.

import type { ParadaCorrida } from '../datos/tipos';
import type { CondicionAutomatizada } from '../datos/tipos';

export function partesAutomatizadas(texto: string): CondicionAutomatizada {
  const t = (texto ?? '').toLowerCase();
  const salida: CondicionAutomatizada = { iteraciones: null, tiempo: null, llamadas: null, resto: '', automatizada: false };
  let resto = t;
  let m = /(\d+)\s*iteraci\w*/.exec(t);
  if (m) {
    salida.iteraciones = Number(m[1]);
    resto = resto.replace(m[0], ' ');
  }
  m = /(\d+(?:[.,]\d+)?)\s*(min\b|minutos?|horas?|h\b|dias?|días?)/.exec(t);
  if (m) {
    const u = m[2]!;
    salida.tiempo = `${m[1]} ${u.startsWith('min') ? 'min' : u === 'h' || u.startsWith('hora') ? 'h' : 'd'}`;
    resto = resto.replace(m[0], ' ');
  }
  m = /(\d+)\s*llamadas?/.exec(t);
  if (m) {
    salida.llamadas = Number(m[1]);
    resto = resto.replace(m[0], ' ');
  }
  // El resto se conserva tal como lo escribio la persona: solo se limpian los
  // conectores sueltos de los bordes y los espacios dobles.
  resto = resto.replace(/\s+/g, ' ').replace(/^[\s,;.]+|[\s,;.]+$/g, '');
  resto = resto.replace(/^(o|y|u|e|,|;)\s+/, '').replace(/\s+(o|y|u|e)$/, '').replace(/^[\s,;.]+|[\s,;.]+$/g, '');
  salida.resto = resto.length >= 4 ? resto : '';
  salida.automatizada = salida.iteraciones !== null || salida.tiempo !== null || salida.llamadas !== null;
  return salida;
}

export function textoAutomatizacion(p: CondicionAutomatizada): string {
  const medibles: string[] = [];
  if (p.iteraciones !== null) medibles.push(`${p.iteraciones} iteraciones`);
  if (p.tiempo) medibles.push(`${p.tiempo} de corrida`);
  if (p.llamadas !== null) medibles.push(`${p.llamadas} llamadas`);
  if (medibles.length === 0) return 'Rosa no puede medir esta condición: la corrida sigue hasta que la detengas o hasta agotar el presupuesto de la misión.';
  let frase = `Rosa para sola al llegar a ${medibles.join(' o ')} (y al agotar el presupuesto de la mision)`;
  if (p.resto) frase += `. El resto ("${p.resto.slice(0, 80)}") lo decides tu con el boton de detener`;
  return `${frase}.`;
}


// ---------------------------------------------------------------------------
// La parada propia de una corrida (15 de septiembre de 2026): lo que la
// persona fija al crearla (horas, iteraciones, llamadas, texto) y que la
// detiene con lo que llegue primero, además de la condición de parada de la
// investigación. Espejo de rosa/parada.py (normalizar_parada, resumen_parada).
// ---------------------------------------------------------------------------

const LIMITES_PARADA: Record<'horas' | 'iteraciones' | 'llamadas' | 'cuantas' | 'sinCambio', [number, number]> = { horas: [1 / 60, 24 * 14], iteraciones: [1, 200], llamadas: [10, 100_000], cuantas: [1, 50], sinCambio: [1, 20] };
export const NIVELES_OBJETIVO = ['baja', 'moderada', 'alta'] as const;

export interface ParadaBorrador {
  horas: string;
  /** Unidad del valor escrito; se convierte a horas para el servidor. */
  unidadTiempo?: 'minutos' | 'horas' | 'dias';
  iteraciones: string;
  llamadas: string;
  texto: string;
  certeza: '' | 'baja' | 'moderada' | 'alta';
  cuantas: string;
  sinCambio: string;
}

export const BORRADOR_VACIO: ParadaBorrador = { horas: '', iteraciones: '', llamadas: '', texto: '', certeza: '', cuantas: '', sinCambio: '' };

function numeroParada(v: string, clave: 'horas' | 'iteraciones' | 'llamadas' | 'cuantas' | 'sinCambio'): number | null {
  const n = Number(v.replace(',', '.'));
  if (!v.trim() || !Number.isFinite(n) || n <= 0) return null;
  const [minimo, maximo] = LIMITES_PARADA[clave];
  const acotado = Math.max(minimo, Math.min(maximo, n));
  return clave === 'horas' ? acotado : Math.floor(acotado);
}

/** Del formulario a la parada que viaja al servidor; null si no se fijó nada. */
export function normalizarParada(b: ParadaBorrador): ParadaCorrida | null {
  const certeza = (NIVELES_OBJETIVO as readonly string[]).includes(b.certeza) ? (b.certeza as 'baja' | 'moderada' | 'alta') : null;
  const p: ParadaCorrida = {
    horas: numeroParada(b.horas.trim() ? String(Number(b.horas.replace(',', '.')) * (b.unidadTiempo === 'minutos' ? 1 / 60 : b.unidadTiempo === 'dias' ? 24 : 1)) : '', 'horas'),
    iteraciones: numeroParada(b.iteraciones, 'iteraciones'),
    llamadas: numeroParada(b.llamadas, 'llamadas'),
    texto: b.texto.trim().slice(0, 300),
    certeza,
    cuantas: certeza ? (numeroParada(b.cuantas, 'cuantas') ?? 1) : null,
    sinCambio: numeroParada(b.sinCambio, 'sinCambio'),
  };
  if (p.horas === null && p.iteraciones === null && p.llamadas === null && !p.texto && p.certeza === null && p.sinCambio === null) return null;
  return p;
}

export function borradorDe(p: ParadaCorrida | null | undefined): ParadaBorrador {
  if (!p) return { ...BORRADOR_VACIO };
  return {
    horas: p.horas === null ? '' : String(p.horas),
    iteraciones: p.iteraciones === null ? '' : String(p.iteraciones),
    llamadas: p.llamadas === null ? '' : String(p.llamadas),
    texto: p.texto ?? '',
    certeza: p.certeza ?? '',
    cuantas: p.cuantas == null ? '' : String(p.cuantas),
    sinCambio: p.sinCambio == null ? '' : String(p.sinCambio),
  };
}

function horasTexto(h: number): string {
  if (h >= 24 && Number.isInteger(h / 24)) return `${h / 24} ${h === 24 ? 'día' : 'días'}`;
  if (h < 1) {
    const m = Math.round(h * 60);
    return m === 1 ? '1 minuto' : `${m} minutos`;
  }
  return `${h} ${h === 1 ? 'hora' : 'horas'}`;
}

/** "2 horas o 6 iteraciones, lo que llegue primero". Vacío sin parada. */
export function resumenParada(p: ParadaCorrida | null | undefined): string {
  if (!p) return '';
  const partes: string[] = [];
  if (p.horas) partes.push(horasTexto(p.horas));
  if (p.iteraciones) partes.push(`${p.iteraciones} ${p.iteraciones === 1 ? 'iteración' : 'iteraciones'}`);
  if (p.llamadas) partes.push(`${p.llamadas} llamadas al modelo`);
  if (p.certeza) {
    const n = p.cuantas ?? 1;
    partes.push(n > 1 ? `${n} hipótesis en certeza ${p.certeza}` : `una hipótesis en certeza ${p.certeza}`);
  }
  if (p.sinCambio) partes.push(`${p.sinCambio} ${p.sinCambio === 1 ? 'iteración' : 'iteraciones'} sin avance`);
  if (p.texto) partes.push(`«${p.texto}»`);
  if (partes.length === 0) return '';
  if (partes.length === 1) return partes[0] ?? '';
  const ultima = partes[partes.length - 1] ?? '';
  return `${partes.slice(0, -1).join(', ')} o ${ultima}, lo que llegue primero`;
}

File: rosa/parada.py
"""Que parte de una condicion de parada puede automatizar Rosa.

La persona escribe la condicion en lenguaje corriente ("3 iteraciones o
cuando el modelo de mundo deje de cambiar"). El bucle solo puede medir tres
cosas: un numero de iteraciones, un tiempo de corrida y un numero de
llamadas al modelo (mas el presupuesto de la mision en dolares y horas). Lo
demas lo decide la persona con el boton de detener. Esta funcion dice, al
escribirla, que quedo automatizado y que no, para que la interfaz lo
muestre en vez de prometer mas de lo que hace. Espejo en
`frontend/src/lib/parada.ts`.
"""

from __future__ import annotations

import re
from typing import Any

_ITERACIONES = re.compile(r"(\d+)\s*iteraci\w*")
_TIEMPO = re.compile(r"(\d+(?:[.,]\d+)?)\s*(min\b|minutos?|horas?|h\b|dias?|días?)")
_LLAMADAS = re.compile(r"(\d+)\s*llamadas?")


def partes_automatizadas(texto: str) -> dict[str, Any]:
    """{iteraciones, tiempo, llamadas, resto, automatizada}. `resto` es lo que
    queda de la condición una vez quitadas las partes medibles; si no está
    vacío, esa parte la decide una persona."""
    t = (texto or "").lower()
    salida: dict[str, Any] = {"iteraciones": None, "tiempo": None, "llamadas": None, "resto": "", "automatizada": False}
    resto = t
    m = _ITERACIONES.search(t)
    if m:
        salida["iteraciones"] = int(m.group(1))
        resto = resto.replace(m.group(0), " ")
    m = _TIEMPO.search(t)
    if m:
        unidad = m.group(2)
        salida["tiempo"] = f"{m.group(1)} {'min' if unidad.startswith('min') else 'h' if unidad in ('h', 'hora') or unidad.startswith('hora') else 'd'}"
        resto = resto.replace(m.group(0), " ")
    m = _LLAMADAS.search(t)
    if m:
        salida["llamadas"] = int(m.group(1))
        resto = resto.replace(m.group(0), " ")
    # El resto se conserva tal como lo escribio la persona: solo se limpian los
    # conectores sueltos de los bordes ("o", ", o", "y") y los espacios dobles.
    resto = re.sub(r"\s+", " ", resto).strip(" ,;.")
    resto = re.sub(r"^(o|y|u|e|,|;)\s+", "", resto).strip(" ,;.")
    resto = re.sub(r"\s+(o|y|u|e)$", "", resto).strip(" ,;.")
    salida["resto"] = resto if len(resto) >= 4 else ""
    salida["automatizada"] = any(salida[k] is not None for k in ("iteraciones", "tiempo", "llamadas"))
    return salida


def texto_automatizacion(partes: dict[str, Any]) -> str:
    """Una frase para la interfaz y el informe."""
    medibles = []
    if partes.get("iteraciones") is not None:
        medibles.append(f"{partes['iteraciones']} iteraciones")
    if partes.get("tiempo"):
        medibles.append(f"{partes['tiempo']} de corrida")
    if partes.get("llamadas") is not None:
        medibles.append(f"{partes['llamadas']} llamadas")
    if not medibles:
        return "Rosa no puede medir esta condición: la corrida sigue hasta que la detengas o hasta agotar el presupuesto de la misión."
    frase = "Rosa para sola al llegar a " + " o ".join(medibles) + " (y al agotar el presupuesto de la misión)"
    if partes.get("resto"):
        frase += f'. El resto ("{partes["resto"][:80]}") lo decides tu con el botón de detener'
    return frase + "."


# ---------------------------------------------------------------------------
# Parada por corrida (15 de septiembre de 2026): al crear una corrida nueva la
# persona puede fijar cuánto debe durar como mucho (horas), cuántas
# iteraciones, cuántas llamadas, o una condición en texto. Se detiene con lo
# que llegue primero, y la condición de parada de la investigación sigue
# valiendo además. Todo opcional: sin nada, la corrida se comporta como antes.
# ---------------------------------------------------------------------------

LIMITES_PARADA = {"horas": (0.05, 24 * 14), "iteraciones": (1, 200), "llamadas": (10, 100_000), "cuantas": (1, 50), "sinCambio": (1, 20)}
NIVELES_OBJETIVO = ("baja", "moderada", "alta")


def normalizar_parada(d: Any) -> dict[str, Any] | None:
    """La parada tal como la guarda la corrida: números acotados o None, texto
    recortado. None si no hay ninguna condición."""
    if not isinstance(d, dict):
        return None
    salida: dict[str, Any] = {"horas": None, "iteraciones": None, "llamadas": None, "texto": "", "certeza": None, "cuantas": None, "sinCambio": None}
    for clave, (minimo, maximo) in LIMITES_PARADA.items():
        v = d.get(clave)
        if v in (None, "", False):
            continue
        try:
            n = float(str(v).replace(",", "."))
        except ValueError:
            continue
        if n <= 0:
            continue
        n = max(minimo, min(maximo, n))
        salida[clave] = round(n, 2) if clave == "horas" else int(n)
    texto = d.get("texto")
    if isinstance(texto, str) and texto.strip():
        salida["texto"] = texto.strip()[:300]
    certeza = d.get("certeza")
    if isinstance(certeza, str) and certeza in NIVELES_OBJETIVO:
        salida["certeza"] = certeza
        salida["cuantas"] = salida["cuantas"] or 1
    else:
        salida["cuantas"] = None  # cuantas solo tiene sentido con un nivel de certeza
    if salida["horas"] is None and salida["iteraciones"] is None and salida["llamadas"] is None and not salida["texto"] and salida["certeza"] is None and salida["sinCambio"] is None:
        return None
    return salida


def _horas_texto(h: float) -> str:
    if h < 1:
        m = int(round(h * 60))
        return f"{m} minutos" if m != 1 else "1 minuto"
    return f"{h:g} {'hora' if h == 1 else 'horas'}"


def resumen_parada(p: dict[str, Any] | None) -> str:
    """Una línea para la persona y para el plan: '2 horas o 6 iteraciones, lo
    que llegue primero'."""
    if not p:
        return ""
    partes = []
    if p.get("horas"):
        partes.append(_horas_texto(float(p["horas"])))
    if p.get("iteraciones"):
        partes.append(f"{p['iteraciones']} {'iteración' if p['iteraciones'] == 1 else 'iteraciones'}")
    if p.get("llamadas"):
        partes.append(f"{p['llamadas']} llamadas al modelo")
    if p.get("certeza"):
        n = int(p.get("cuantas") or 1)
        partes.append(f"{n} {'hipótesis' } en certeza {str(p['certeza']).replace('_', ' ')}" if n > 1 else f"una hipótesis en certeza {str(p['certeza']).replace('_', ' ')}")
    if p.get("sinCambio"):
        partes.append(f"{p['sinCambio']} {'iteración' if p['sinCambio'] == 1 else 'iteraciones'} sin avance")
    if p.get("texto"):
        partes.append(f"«{p['texto']}»")
    if not partes:
        return ""
    return partes[0] if len(partes) == 1 else ", ".join(partes[:-1]) + " o " + partes[-1] + ", lo que llegue primero"


def texto_condicion(inv: dict[str, Any], c: dict[str, Any] | None) -> str:
    """La condición completa que ve el planificador: la de esta corrida (si la
    hay) y la de la investigación, que siempre vale."""
    propia = resumen_parada((c or {}).get("parada"))
    base = (inv.get("condicionParada") or "").strip() or "Sin condición de parada declarada"
    if not propia:
        return base
    return f"Esta corrida: como mucho {propia}. Además sigue valiendo la condición de la investigación: {base}"
